yoyodyne/models/modules/transformer.py

```diff
@@ -184,8 +223,248 @@ def name(self) -> str:
+                features_memory_mask,
+                features_memory_mask,
+                # FIXME Introduced in torch 2.0
+                # memory_is_causal,
+            )
+            # TODO: Do we want a nonlinear activation?
+            feature_attention = self.features_linear(feature_attention)
+            x = torch.cat([symbol_attention, feature_attention], dim=2)
+            x = x + self._ff_block(self.norm3(x))
+        else:
+            x = self.norm1(
+                x
+                + self._sa_block(
+                    x,
+                    target_mask,
+                    target_key_padding_mask,
+                    # FIXME: Introduced in torch 2.0
+                    # is_causal=target_is_causal
+                )
+            )
+            symbol_attention = self._mha_block(
+                x,
+                memory,
+                memory_mask,
+                memory_key_padding_mask,
+                # FIXME Introduced in torch 2.0
+                # memory_is_causal,
+            )
+            # TODO: Do we want a nonlinear activation?
+            symbol_attention = self.symbols_linear(symbol_attention)
+            feature_attention = self._features_mha_block(
+                x,
+                features_memory,
+                features_memory_mask,
+                features_memory_mask,
+                # FIXME Introduced in torch 2.0
+                # memory_is_causal,
+            )
+            # TODO: Do we want a nonlinear activation?
+            feature_attention = self.features_linear(feature_attention)
+            x = x + torch.cat([symbol_attention, feature_attention], dim=2)
+            x = self.norm2(x)
+            x = self.norm3(x + self._ff_block(x))
+        return x
+
+    def _features_mha_block(
+        self,
+        x: torch.Tensor,
+        mem: torch.Tensor,
+        attn_mask: Optional[torch.Tensor],
+        key_padding_mask: Optional[torch.Tensor],
+        # FIXME: Introduced in torch 2.0
+        # is_causal: bool = False,
+    ) -> torch.Tensor:
+        """Runs the multihead attention block that attends to features.
+
+        Args:
+            x (torch.Tensor): The `query` tensor, i.e. the previous decoded
+                embeddings.
+            mem (torch.Tensor): The `keys` and `values`, i.e. the encoded
+                features.
+            attn_mask (torch.Tensor, optional): the mask for the features.
+            key_padding_mask (torch.Tensor, optional): the mask for the
+                feature keys per batch.
+
+        Returns:
+            torch.Tensor: Concatenated attention head tensors.
+        """
+        x = self.feature_multihead_attn(
+            x,
+            mem,
+            mem,
+            attn_mask=attn_mask,
+            key_padding_mask=key_padding_mask,
+            # FIXME: Introduced in torch 2.0
+            # is_causal=is_causal,
+            need_weights=False,
+        )[0]
+        return self.dropout2(x)
```

Comment:
Another dumb question: where is the `dropout2` object or method implemented? Did I just miss it?

Reply:
I believe this is in the torch `TransformerDecoderLayer` that I inherit from.